### ISSUE TITLE
Question about the gan training

### ISSUE THREAD

**Issue by DRJYYDS** [hlzhang109/DDG#5]
Hi could you describe how you train the Gan in the training process.
I haven’t see the description in your paper's algorithm.
And your code seems too complex to read due to no comments.
It's confusing that whai is dis_update , gen_update, gan_update:)
Looking forward to your reply!

**Comment by DRJYYDS** [comment]
![image](https://user-images.githubusercontent.com/102657334/160792477-02592aa1-39ae-4dfa-9049-f5f554effb44.png)
here is the traning code. Could you explain it into more details?

**Comment by DRJYYDS** [comment]
![image](https://user-images.githubusercontent.com/102657334/160792682-8c394eac-a000-4d0a-b522-3dbe775e8a53.png)
Here is the training code. Could you guys explain it more in detailed?
Or just give me a description like what you guys do in the Algorithm1 which shown in paper.

**Comment by yfzhang114** [comment]
Please refer to https://github.com/NVlabs/MUNIT for a detailed explanation of the GAN,

**Comment by DRJYYDS** [comment]
> Please refer to https://github.com/NVlabs/MUNIT for a detailed explanation of the GAN,

This is not my question. Actually, I know Gans well. The question for me is how you using the gan in your algorithm, what is the final loss for you method, these are the things you didn't mention in your paper, and your code is kind of confusing due to no comments, can you reply to my question seriously?

**Comment by yfzhang114** [comment]
Both dis_update and gan_update are used for training the GAN, on which dis_update is used to train the discriminator, and gan_update is used to train the generator. The additional cycle-consistent loss, gan loss are all used for training the GAN and we have mentioned these additional losses in Section 4.

In gen_update, we 
1. calculate the reconstruction loss.
2. calculate the empirical loss
3. update the primal-dual parameter
![image](https://user-images.githubusercontent.com/53161950/161170632-6cb4e80b-9d98-47b9-bd01-7f8b92ae0ab3.png)

**Comment by DRJYYDS** [comment]
> 

Thanks for your explanation. But there is still some confusion for me. Your paper did mention the loss of Gan, but you didn't talk about the pre-train step and this make reader think your algorithm is 1stage in training. If you guys consider update the discussion about pre-train step in paper, I would pretty appreciate it, also I think this would make your paper be more attractive and make more important contributions to the community.

**Comment by DRJYYDS** [comment]
> Both dis_update and gan_update are used for training the GAN, on which dis_update is used to train the discriminator, and gan_update is used to train the generator. The additional cycle-consistent loss, gan loss are all used for training the GAN and we have mentioned these additional losses in Section 4.
> 
> In gen_update, we
> 
> 1. calculate the reconstruction loss.
> 2. calculate the empirical loss
> 3. update the primal-dual parameter
>    ![image](https://user-images.githubusercontent.com/53161950/161170632-6cb4e80b-9d98-47b9-bd01-7f8b92ae0ab3.png)


Thanks for your explanation. But there is still some confusion for me. Your paper did mention the loss of Gan, but you didn't talk about the pre-train step and this make reader think your algorithm is 1stage in training. If you guys consider update the discussion about pre-train step in paper, I would pretty appreciate it, also I think this would make your paper be more attractive and make more important contributions to the community.

**Comment by yfzhang114** [comment]
The pre-train step is not necessary and is just for better visualization results.

**Comment by gkyustc** [comment]
> Both dis_update and gan_update are used for training the GAN, on which dis_update is used to train the discriminator, and gan_update is used to train the generator. The additional cycle-consistent loss, gan loss are all used for training the GAN and we have mentioned these additional losses in Section 4.
> 
> In gen_update, we
> 
> 1. calculate the reconstruction loss.
> 2. calculate the empirical loss
> 3. update the primal-dual parameter
>    ![image](https://user-images.githubusercontent.com/53161950/161170632-6cb4e80b-9d98-47b9-bd01-7f8b92ae0ab3.png)

hi,
Thank you for your impressive work. 
I notice that, in stage 2 of this image, the f_a_recon features in mnist dataset and the same features in other datasets are from different generated images(x_ba for mnist and x_ab for other datasets).  Do you have any explanation or insight for this selection?
